enterprise/server/quota/quota_manager.go

```diff
@@ -467,70 +480,121 @@ func (qm *QuotaManager) ApplyBucket(ctx context.Context, req *qpb.ApplyBucketReq
 	if req.GetNamespace() == "" {
-		return nil, status.InvalidArgumentError("namespace cannot empty")
+		return nil, status.InvalidArgumentError("namespace cannot be empty")
+	}
+
+	if req.GetAddBucket() == nil && req.GetUpdateBucket() == nil && req.GetRemoveBucket() == "" {
+		return nil, status.InvalidArgumentError("one of add_bucket, update_bucket and remove_bucket should be set")
 	}
 
 	if req.GetAddBucket() != nil {
-		return qm.addBucket(ctx, req.GetNamespace(), req.GetAddBucket())
+		if err := qm.addBucket(ctx, req.GetNamespace(), req.GetAddBucket()); err != nil {
+			return &qpb.ModifyNamespaceResponse{}, err
+		}
 	}
 
 	if req.GetUpdateBucket() != nil {
-		return qm.updateBucket(ctx, req.GetNamespace(), req.GetUpdateBucket())
+		if err := qm.updateBucket(ctx, req.GetNamespace(), req.GetUpdateBucket()); err != nil {
+			return &qpb.ModifyNamespaceResponse{}, err
+		}
 	}
 
 	if req.GetRemoveBucket() != "" {
-		return qm.removeBucket(ctx, req.GetNamespace(), req.GetRemoveBucket())
+		if err := qm.removeBucket(ctx, req.GetNamespace(), req.GetRemoveBucket()); err != nil {
+			return &qpb.ModifyNamespaceResponse{}, err
+		}
 	}
 
-	return nil, status.InvalidArgumentError("one of add_bucket, update_bucket and remove_bucket should be set")
+	qm.notifyUpdates()
+
+	return &qpb.ModifyNamespaceResponse{}, nil
 }
 
-func (qm *QuotaManager) addBucket(ctx context.Context, namespace string, bucket *qpb.Bucket) (*qpb.ModifyNamespaceResponse, error) {
+func (qm *QuotaManager) addBucket(ctx context.Context, namespace string, bucket *qpb.Bucket) error {
 	if err := validateBucket(bucket); err != nil {
-		return nil, status.InvalidArgumentErrorf("invalid add_bucket: %s", err)
+		return status.InvalidArgumentErrorf("invalid add_bucket: %s", err)
 	}
 	row := bucketToRow(namespace, bucket)
 
 	err := qm.env.GetDBHandle().DB(ctx).Create(&row).Error
 	if err != nil {
-		return nil, err
+		return err
 	}
-	return &qpb.ModifyNamespaceResponse{}, nil
+	return nil
 }
 
-func (qm *QuotaManager) updateBucket(ctx context.Context, namespace string, bucket *qpb.Bucket) (*qpb.ModifyNamespaceResponse, error) {
+func (qm *QuotaManager) updateBucket(ctx context.Context, namespace string, bucket *qpb.Bucket) error {
 	if err := validateBucket(bucket); err != nil {
-		return nil, status.InvalidArgumentErrorf("invalid update_bucket: %s", err)
+		return status.InvalidArgumentErrorf("invalid update_bucket: %s", err)
 	}
 	bucketRow := bucketToRow(namespace, bucket)
 
 	db := qm.env.GetDBHandle().DB(ctx)
 	res := db.Model(bucketRow).Where("namespace = ? AND name= ?", bucketRow.Namespace, bucketRow.Name).Updates(bucketRow)
 
 	if res.Error != nil {
-		return nil, res.Error
+		return res.Error
 	}
 	if res.RowsAffected == 0 {
-		return nil, status.InvalidArgumentErrorf("bucket %s doesn't exist", bucket.GetName())
+		return status.InvalidArgumentErrorf("bucket %s doesn't exist", bucket.GetName())
```

Comment:
usually i think it's nice to quote strings that are use-supplied when you log them, with %q instead of %s